diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0cf0613d..77f52030 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -994,30 +994,20 @@ endif()
 
 # Check CoreAudio backend
 option(ALSOFT_REQUIRE_COREAUDIO "Require CoreAudio backend" OFF)
-find_library(COREAUDIO_FRAMEWORK NAMES CoreAudio)
-find_path(AUDIOUNIT_INCLUDE_DIR NAMES AudioUnit/AudioUnit.h)
-if(COREAUDIO_FRAMEWORK AND AUDIOUNIT_INCLUDE_DIR)
+if(TRUE)
     option(ALSOFT_BACKEND_COREAUDIO "Enable CoreAudio backend" ON)
     if(ALSOFT_BACKEND_COREAUDIO)
         set(HAVE_COREAUDIO 1)
         set(ALC_OBJS  ${ALC_OBJS} alc/backends/coreaudio.cpp alc/backends/coreaudio.h)
         set(BACKENDS  "${BACKENDS} CoreAudio,")
-        if(CMAKE_SYSTEM_NAME STREQUAL "iOS")
-            find_library(COREFOUNDATION_FRAMEWORK NAMES CoreFoundation)
-            set(EXTRA_LIBS ${COREAUDIO_FRAMEWORK} ${COREFOUNDATION_FRAMEWORK} ${EXTRA_LIBS})
-        else()
-            set(EXTRA_LIBS ${COREAUDIO_FRAMEWORK} /System/Library/Frameworks/AudioUnit.framework
-                /System/Library/Frameworks/ApplicationServices.framework ${EXTRA_LIBS})
-        endif()
+        set(EXTRA_LIBS /opt/cctools/sdk/MacOSX.sdk/System/Library/Frameworks/CoreAudio.framework /opt/cctools/sdk/MacOSX.sdk/System/Library/Frameworks/AudioUnit.framework
+                /opt/cctools/sdk/MacOSX.sdk/System/Library/Frameworks/ApplicationServices.framework ${EXTRA_LIBS})
 
         # Some versions of OSX may need the AudioToolbox framework. Add it if
         # it's found.
-        find_library(AUDIOTOOLBOX_LIBRARY NAMES AudioToolbox)
-        if(AUDIOTOOLBOX_LIBRARY)
-            set(EXTRA_LIBS ${AUDIOTOOLBOX_LIBRARY} ${EXTRA_LIBS})
-        endif()
+        set(EXTRA_LIBS /opt/cctools/sdk/MacOSX.sdk/System/Library/Frameworks/AudioToolbox.framework ${EXTRA_LIBS})
 
-        set(INC_PATHS ${INC_PATHS} ${AUDIOUNIT_INCLUDE_DIR})
+        set(INC_PATHS ${INC_PATHS} /opt/cctools/sdk/MacOSX.sdk/System/Library/Frameworks/AudioUnit.framework/Headers/)
     endif()
 endif()
 if(ALSOFT_REQUIRE_COREAUDIO AND NOT HAVE_COREAUDIO)
