diff --git a/CMakeLists.txt b/CMakeLists.txt
index 55644b0..7cd8fd8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1070,31 +1070,20 @@ endif()
 option(ALSOFT_BACKEND_COREAUDIO "Enable CoreAudio backend" ON)
 option(ALSOFT_REQUIRE_COREAUDIO "Require CoreAudio backend" OFF)
 if(ALSOFT_BACKEND_COREAUDIO)
-    find_library(COREAUDIO_FRAMEWORK NAMES CoreAudio)
-    find_path(AUDIOUNIT_INCLUDE_DIR NAMES AudioUnit/AudioUnit.h)
-    if(COREAUDIO_FRAMEWORK AND AUDIOUNIT_INCLUDE_DIR)
+    if(TRUE)
         set(HAVE_COREAUDIO 1)
         set(ALC_OBJS  ${ALC_OBJS} alc/backends/coreaudio.cpp alc/backends/coreaudio.h)
         set(BACKENDS  "${BACKENDS} CoreAudio,")
 
         set(EXTRA_LIBS -Wl,-framework,CoreAudio ${EXTRA_LIBS})
-        if(CMAKE_SYSTEM_NAME MATCHES "^(iOS|tvOS)$")
-            find_library(COREFOUNDATION_FRAMEWORK NAMES CoreFoundation)
-            if(COREFOUNDATION_FRAMEWORK)
-                set(EXTRA_LIBS -Wl,-framework,CoreFoundation ${EXTRA_LIBS})
-            endif()
-        else()
-            set(EXTRA_LIBS -Wl,-framework,AudioUnit,-framework,ApplicationServices ${EXTRA_LIBS})
-        endif()
+        set(EXTRA_LIBS /opt/cctools/sdk/MacOSX.sdk/System/Library/Frameworks/CoreAudio.framework /opt/cctools/sdk/MacOSX.sdk/System/Library/Frameworks/AudioUnit.framework
+                /opt/cctools/sdk/MacOSX.sdk/System/Library/Frameworks/ApplicationServices.framework ${EXTRA_LIBS})
 
         # Some versions of OSX may need the AudioToolbox framework. Add it if
         # it's found.
-        find_library(AUDIOTOOLBOX_LIBRARY NAMES AudioToolbox)
-        if(AUDIOTOOLBOX_LIBRARY)
-            set(EXTRA_LIBS -Wl,-framework,AudioToolbox ${EXTRA_LIBS})
-        endif()
+        set(EXTRA_LIBS /opt/cctools/sdk/MacOSX.sdk/System/Library/Frameworks/AudioToolbox.framework ${EXTRA_LIBS})
 
-        set(INC_PATHS ${INC_PATHS} ${AUDIOUNIT_INCLUDE_DIR})
+        set(INC_PATHS ${INC_PATHS} /opt/cctools/sdk/MacOSX.sdk/System/Library/Frameworks/AudioUnit.framework/Headers/)
     endif()
 endif()
 if(ALSOFT_REQUIRE_COREAUDIO AND NOT HAVE_COREAUDIO)
