diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0cf0613..b9a1cd9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -205,10 +205,10 @@ endif()
 
 if(MSVC)
     set(CPP_DEFS ${CPP_DEFS} _CRT_SECURE_NO_WARNINGS NOMINMAX)
-    check_cxx_compiler_flag(/permissive- HAVE_PERMISSIVE_SWITCH)
-    if(HAVE_PERMISSIVE_SWITCH)
-        set(C_FLAGS ${C_FLAGS} $<$<COMPILE_LANGUAGE:CXX>:/permissive->)
-    endif()
+    # check_cxx_compiler_flag(/permissive- HAVE_PERMISSIVE_SWITCH)
+    # if(HAVE_PERMISSIVE_SWITCH)
+        # set(C_FLAGS ${C_FLAGS} $<$<COMPILE_LANGUAGE:CXX>:/permissive->)
+    # endif()
     set(C_FLAGS ${C_FLAGS} /W4 /w14640 /wd4065 /wd4268 /wd4324 /wd5030)
     # Remove /W3, which is added by default, since we set /W4. Some build
     # generators with MSVC complain about both /W3 and /W4 being specified.
@@ -435,7 +435,7 @@ set(FPMATH_SET "0")
 if(CMAKE_SIZEOF_VOID_P MATCHES "4" AND HAVE_SSE2)
     option(ALSOFT_ENABLE_SSE2_CODEGEN "Enable SSE2 code generation instead of x87 for 32-bit targets." TRUE)
     if(ALSOFT_ENABLE_SSE2_CODEGEN)
-        if(SSE2_SWITCH OR NOT MSVC)
+        if(NOT MSVC)
             check_c_compiler_flag("${SSE2_SWITCH} -mfpmath=sse" HAVE_MFPMATH_SSE_2)
             if(HAVE_MFPMATH_SSE_2)
                 set(SSE_FLAGS ${SSE_FLAGS} ${SSE2_SWITCH} -mfpmath=sse)
