diff --git a/cmake/Findfutimens.cmake b/cmake/Findfutimens.cmake
index 3449c9d54..93a1b34da 100644
--- a/cmake/Findfutimens.cmake
+++ b/cmake/Findfutimens.cmake
@@ -6,9 +6,13 @@ if(APPLE)
 	# that doesn't have futimens.  Instead we need to enable warnings
 	# as errors, then check for the symbol existing in `sys/stat.h`,
 	# then reset warnings as errors.
+	list(APPEND CMAKE_REQUIRED_FLAGS -Werror=unguarded-availability)
 	enable_warnings(error)
 	check_symbol_exists(futimens sys/stat.h HAVE_FUTIMENS)
+	check_symbol_exists(getentropy unistd.h HAVE_RAND_GETENTROPY)
 	disable_warnings(error)
+	list(REMOVE_ITEM CMAKE_REQUIRED_FLAGS -Werror=unguarded-availability)
 else()
 	check_function_exists(futimens HAVE_FUTIMENS)
+	check_symbol_exists(getentropy unistd.h HAVE_RAND_GETENTROPY)
 endif()
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index ed3f4a514..14f1f608d 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -84,7 +84,9 @@ check_prototype_definition_safe(qsort_s
 
 # random / entropy data
 
-check_symbol_exists(getentropy unistd.h GIT_RAND_GETENTROPY)
+if(HAVE_RAND_GETENTROPY)
+	set(GIT_RAND_GETENTROPY 1)
+endif()
 check_symbol_exists(getloadavg stdlib.h GIT_RAND_GETLOADAVG)
 
 # poll
