diff -Naru v1.8.0-1cd8a38f70.clean/src/util/util.h v1.8.0-1cd8a38f70/src/util/util.h
--- v1.8.0-1cd8a38f70.clean/src/util/util.h	2026-03-23 09:49:32.697416300 +0100
+++ v1.8.0-1cd8a38f70/src/util/util.h	2026-03-23 09:49:37.036683000 +0100
@@ -287,9 +287,26 @@
 
 GIT_INLINE(uint64_t) git_time_monotonic(void)
 {
+#if _WIN32_WINNT >= 0x0600
 	/* GetTickCount64 returns the number of milliseconds that have
 	 * elapsed since the system was started. */
 	return GetTickCount64();
+#else
+	/* We need the initial tick count to detect if the tick
+	 * count has rolled over. */
+	static DWORD initial_tick_count = 0;
+
+	/* GetTickCount returns the number of milliseconds that have
+	 * elapsed since the system was started. */
+	DWORD count = GetTickCount();
+	if(initial_tick_count == 0) {
+		initial_tick_count = count;
+	} else if(count < initial_tick_count) {
+		/* The tick count has rolled over - adjust for it. */
+		count = (0xFFFFFFFFu - initial_tick_count) + count;
+	}
+	return count;
+#endif
 }
 
 #elif __APPLE__
