From 477a60b95b6cfa435cafbd511e5954077847845b Mon Sep 17 00:00:00 2001
From: Edoardo Lolletti <edoardo762@gmail.com>
Date: Sat, 19 Feb 2022 16:44:22 +0100
Subject: [PATCH] Update flags for WideCharToMultiByte to work on winxp

---
 src/util/win32/utf-conv.c | 19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/src/util/win32/utf-conv.c b/src/util/win32/utf-conv.c
index 4bde3023a..b2d66b25e 100644
--- a/src/util/win32/utf-conv.c
+++ b/src/util/win32/utf-conv.c
@@ -36,6 +36,19 @@ int git__utf8_to_16(wchar_t *dest, size_t dest_size, const char *src)
 	return len;
 }
 
+GIT_INLINE(DWORD) get_wc_flags(void) {
+    static char inited = 0;
+    static DWORD flags;
+
+    /* Invalid code point check supported on Vista+ only */
+    if(!inited) {
+		flags = git_has_win32_version(6, 0, 0) ? WC_ERR_INVALID_CHARS : 0;
+		inited = 1;
+    }
+
+    return flags;
+}
+
 /**
  * Converts a wide string to UTF-8.
  *
@@ -51,7 +64,7 @@ int git__utf16_to_8(char *dest, size_t dest_size, const wchar_t *src)
 	/* Length of -1 indicates NULL termination of the input string. Subtract 1 from the result to
 	 * turn 0 into -1 (an error code) and to not count the NULL terminator as part of the string's
 	 * length. WideCharToMultiByte never returns int's minvalue, so underflow is not possible */
-	if ((len = WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, src, -1, dest, (int)dest_size, NULL, NULL) - 1) < 0)
+	if ((len = WideCharToMultiByte(CP_UTF8, get_wc_flags(), src, -1, dest, (int)dest_size, NULL, NULL) - 1) < 0)
 		git__set_errno();
 
 	return len;
@@ -116,7 +129,7 @@ int git__utf16_to_8_alloc(char **dest, const wchar_t *src)
 	*dest = NULL;
 
 	/* Length of -1 indicates NULL termination of the input string */
-	utf8_size = WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, src, -1, NULL, 0, NULL, NULL);
+	utf8_size = WideCharToMultiByte(CP_UTF8, get_wc_flags(), src, -1, NULL, 0, NULL, NULL);
 
 	if (!utf8_size) {
 		git__set_errno();
@@ -130,7 +143,7 @@ int git__utf16_to_8_alloc(char **dest, const wchar_t *src)
 		return -1;
 	}
 
-	utf8_size = WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, src, -1, *dest, utf8_size, NULL, NULL);
+	utf8_size = WideCharToMultiByte(CP_UTF8, get_wc_flags(), src, -1, *dest, utf8_size, NULL, NULL);
 
 	if (!utf8_size) {
 		git__set_errno();
-- 
2.38.1.windows.1

