diff --git a/lib/config-win32.h b/lib/config-win32.h
index 316043d81..a75464d15 100644
--- a/lib/config-win32.h
+++ b/lib/config-win32.h
@@ -543,6 +543,11 @@
 #  endif
 #endif
 
+#undef _WIN32_WINNT
+#define _WIN32_WINNT 0x0501
+#undef WINVER
+#define WINVER 0x0501
+
 /* VS2008 default target settings and minimum build target check. */
 #if defined(_MSC_VER) && (_MSC_VER >= 1500) && (_MSC_VER <= 1600)
 #  ifndef _WIN32_WINNT
diff --git a/lib/curl_threads.h b/lib/curl_threads.h
index 65d1a790c..75f899467 100644
--- a/lib/curl_threads.h
+++ b/lib/curl_threads.h
@@ -23,6 +23,8 @@
  ***************************************************************************/
 #include "curl_setup.h"
 
+#undef _WIN32_WINNT
+
 #if defined(USE_THREADS_POSIX)
 #  define CURL_STDCALL
 #  define curl_mutex_t           pthread_mutex_t
diff --git a/lib/inet_ntop.c b/lib/inet_ntop.c
index 9a5af7f42..f3aef0964 100644
--- a/lib/inet_ntop.c
+++ b/lib/inet_ntop.c
@@ -20,6 +20,8 @@
 
 #include "curl_setup.h"
 
+#undef HAVE_INET_NTOP
+
 #ifndef HAVE_INET_NTOP
 
 #ifdef HAVE_SYS_PARAM_H
diff --git a/lib/inet_ntop.h b/lib/inet_ntop.h
index 9d3f237f3..384228b8a 100644
--- a/lib/inet_ntop.h
+++ b/lib/inet_ntop.h
@@ -24,6 +24,8 @@
 
 #include "curl_setup.h"
 
+#undef HAVE_INET_NTOP
+
 char *Curl_inet_ntop(int af, const void *addr, char *buf, size_t size);
 
 #ifdef HAVE_INET_NTOP
diff --git a/lib/inet_pton.c b/lib/inet_pton.c
index 9c87a0562..c65326caa 100644
--- a/lib/inet_pton.c
+++ b/lib/inet_pton.c
@@ -18,6 +18,7 @@
 
 #include "curl_setup.h"
 
+#undef HAVE_INET_PTON
 #ifndef HAVE_INET_PTON
 
 #ifdef HAVE_SYS_PARAM_H
diff --git a/lib/inet_pton.h b/lib/inet_pton.h
index e695af9c6..ef97245b9 100644
--- a/lib/inet_pton.h
+++ b/lib/inet_pton.h
@@ -26,6 +26,8 @@
 
 int Curl_inet_pton(int, const char *, void *);
 
+#undef HAVE_INET_PTON
+
 #ifdef HAVE_INET_PTON
 #ifdef HAVE_ARPA_INET_H
 #include <arpa/inet.h>
